export/src/types.ts

```
@@ -12,17 +12,24 @@ export type ThemeState = Readonly<{
 }>;
 export type ColorScheme = 'LIGHT_COLOR_SCHEME' | 'DARK_COLOR_SCHEME';
 export type Semester = number;
-export type ClassNo = string; // E.g. "1", "A"
-export type LessonType = string; // E.g. "Lecture", "Tutorial"
 export type ModuleCode = string; // E.g. "CS3216"
+export type LessonType = string; // E.g. "Lecture", "Tutorial"
+export type ClassNo = string; // E.g. "1", "A"
+export type StartTime = string;
+export type DayText = string;
 export type SemTimetableConfig = {
   [moduleCode: ModuleCode]: ModuleLessonConfig;
 };
 export interface ModuleLessonConfig {
   [lessonType: LessonType]: ClassNo;
 }
 export type TaModulesConfig = {
-  [moduleCode: ModuleCode]: [lessonType: LessonType, classNo: ClassNo][];
+  [moduleCode: ModuleCode]: [
+    lessonType: LessonType,
+    classNo: ClassNo,
+    startTime: StartTime,
+    day: DayText,
+  ][];
```

`TaModulesConfig` is starting to get quite big. I'm starting to think that it would be better to store this state differently. With this PR the state would look like:
```
timetable: {
	lessons: {
	  1: {
	     // note: the "CS2103T" keyed object is completely useless since the user is a TA for CS2103T.
	    CS2103T: {
	    	Lecture: "G01"
	   }
	},
	ta: {
		1: {
			CS2103T: [["Lecture", "G01", "0800", "Monday"]]
		}
	}
}
```
Instead, we could have `addTaLesson` actions overwrite the "lessons" keyed object like so:
```
timetable: {
	lessons: {
	  1: {
	    CS2103T: {
				// note: values are now arrays of arrays.
	    	Lecture: [["G01", "0800", "Monday"]]
	   }
	},
	ta: {
		1: ["CS2103T"]
	}
}
```
Note that we would have to change the type of the values for `ModuleLessonConfig` to an array.
Coincidentally (or not), this change causes the state schema for "ta" and "hidden" to match (they are both arrays of `ModuleCode`s.

For readability, we can go even further and replace `["G01", "0800", "Monday"]` with:
```
{
	classNo: "G01",
	startTime: "0800",
	day: "Monday"
}
```
Hence, the resulting state would look like:
```
timetable: {
	lessons: {
	  1: {
	    CS2103T: {
	    	Lecture: [
					{
						classNo: "G01",
						startTime: "0800",
						day: "Monday"
					}
				]
	   }
	},
	ta: {
		1: ["CS2103T"]
	}
}
```

As for the serialized state, it changes from `CS2103T=LEC:G01&ta=CS2103T(LEC:G01:0800:Monday)` to `CS2103T=(LEC:G01:0800:Monday)&ta=CS2103T`.

---

Another thing I'm not sure how to handle is that there are some mods with lessons with identical startTime, endTime and day (e.g. HSI1000). The only thing differentiating the lessons are the weeks and venue. In this case, should we add weeks/venue to the schema as well? I feel that it would be introducing bloat for such a rare scenario though.

---

That's a problem. Can we hold off on the addition of `startTime` and `endTime` and any other distinguishing identifiers for now? Let's just make it possible to add lectures to TA, and figure out what we really need to make this feature possible.

---

Yep, that makes sense to me.

---

@leslieyip02 Regarding the lessons with identical startTime, endTime and day in HSI1000, I am not sure which ones are colliding, it appears that they have different classNo?

---

@leslieyip02 Ah I see I have just noticed that it disappears when modifying the schedule because the dedupe code uses classNo, which is a bug

---

It looks like each module data already has individual lessons in a list, so if we are going the index-as-key route it might be possible to use that. But I am still worried about how stable the list is.

---

In the meantime, I'll also try to look for the edge cases that aren't covered so we can try to figure out what the minimum identifying info is:

- Same classNo, different days: CG1111A
- Same classNo, same day, start time and end time: HSI1000
- Same classNo, same lesson type at different times: CS2103T
- Same classNo, different lesson types at different times: FST2102B

So far, it looks like we need day, start time, end time and lesson type.

---

> But TBH if the lesson indices do changes, the timetable would need to be manually updated by the user anyways since new lesson options are available/existing lessons have changed/been removed.

Yeah... Furthermore, currently things fail silently, so I think it would make sense to notify users when any discrepancies are found. It would be a separate issue though

> We would just need to handle the case where the index is out-of-bounds.

Yeah, that's part of the validation. currently I am handling it naively such that when it finds a non-valid lesson in normal mode it would also try to recover by finding lessons with the same classNo

---

@leslieyip02 For the 4th example, suppose there were 2 classes, 4 lessons

LEC:01 <- currently selected
TUT:01 <- currently selected
LEC:02
TUT:02

Do we need to make it such selecting LEC:02 will also switch the tutorial lesson type to TUT:02?

---

@zwliew From https://github.com/nusmodifications/nusmods/pull/3930#discussion_r1902257170
> When switching from TA to non-TA mode, we should also pick the first lesson for each type and discard the rest.

I think it would be better to take into account the lessons that the user currently have selected in TA mode, it's also not too difficult to implement

Uploading Recording 2025-07-09 192301 (1).mp4…